pkg/engine/datasource/graphql_datasource/graphql_datasource.go

```diff
@@ -1342,31 +1343,33 @@ func (s *Source) compactAndUnNullVariables(input []byte) []byte {
 	}
 
 	removeNullVariables := httpclient.IsInputFlagSet(input, httpclient.UNNULLVARIABLES)
-	variables = s.cleanupVariables(variables, removeNullVariables)
+	variables = s.cleanupVariables(variables, removeNullVariables, undefinedVariables)
 
 	input, _ = jsonparser.Set(input, variables, "body", "variables")
 	return input
 }
 
 // cleanupVariables removes null variables and empty objects from the input if removeNullVariables is true
 // otherwise returns the input as is
-func (s *Source) cleanupVariables(variables []byte, removeNullVariables bool) []byte {
+func (s *Source) cleanupVariables(variables []byte, removeNullVariables bool, undefinedVariables []string) []byte {
 	cp := make([]byte, len(variables))
 	copy(cp, variables)
 
-	if removeNullVariables {
-		// remove null variables from JSON: {"a":null,"b":1} -> {"b":1}
-		err := jsonparser.ObjectEach(variables, func(key []byte, value []byte, dataType jsonparser.ValueType, offset int) error {
-			if dataType == jsonparser.Null {
+	// remove null variables from JSON: {"a":null,"b":1} -> {"b":1}
+	err := jsonparser.ObjectEach(variables, func(key []byte, value []byte, dataType jsonparser.ValueType, offset int) error {
+		if dataType == jsonparser.Null {
+			if removeNullVariables || slices.Contains(undefinedVariables, string(key)) {
```

Comment:
Is it possible to implement this function without casting from `[]byte` to `string`?

Reply:
do you mean `key`?  we need it to be a string to delete it (line 1362)

Reply:
yes, but it should be possible without allocating, as we just read the value
e.g. you can use unsafebytes package to translate byte slice to string without allocation